api/sirsidynix_authentication_provider.py

```diff
@@ -0,0 +1,332 @@
+        self.server_url = self.server_url + (
+            "/" if not self.server_url.endswith("/") else ""
+        )
+
+        self.sirsi_client_id = integration.setting(self.Keys.CLIENT_ID).value
+        self.sirsi_app_id = os.environ.get(
+            Configuration.SIRSI_DYNIX_APP_ID, default=self.DEFAULT_APP_ID
+        )
+        self.sirsi_disallowed_prefixes = integration.setting(
+            self.Keys.DISALLOWED_SUFFIXES
+        ).value_or_default([])
+        self.sirsi_library_id = (
+            ConfigurationSetting.for_library_and_externalintegration(
+                object_session(library), self.Keys.LIBRARY_ID, library, integration
+            ).value
+        )
+        self.sirsi_library_prefix = (
+            ConfigurationSetting.for_library_and_externalintegration(
+                object_session(library), self.Keys.LIBRARY_PREFIX, library, integration
+            ).value
+        )
+
+    def remote_authenticate(self, username: str, password: str) -> PatronData | bool:
+        """Authenticate this user with the remote server."""
+        data = self.api_patron_login(username, password)
+        if not data:
+            return False
+
+        return SirsiDynixPatronData(
+            username=username,
+            authorization_identifier=username,
+            permanent_id=data.get("patronKey"),
+            session_token=data.get("sessionToken"),
+            complete=False,
+        )
+
+    def remote_patron_lookup(
+        self, patron_or_patrondata: Patron | SirsiDynixPatronData
+    ) -> None | SirsiDynixPatronData:
+        """Do a remote patron lookup, this method can only lookup a patron with a patrondata object
+        with a session_token already setup within it.
+        This method also checks all the reasons that a patron may be blocked for.
+        """
+        # We cannot do a remote lookup without a session token
+        if not isinstance(patron_or_patrondata, SirsiDynixPatronData):
+            return None
+        elif not patron_or_patrondata.session_token:
+            return None
+
+        patrondata = patron_or_patrondata
+        # Pull and parse the basic patron information
+        data = self.api_read_patron_data(
+            patron_key=patrondata.permanent_id,
+            session_token=patrondata.session_token,
+        )
+        if not data or "fields" not in data:
+            return None
+
+        fields: dict = data["fields"]
+        patrondata.personal_name = fields.get("displayName")
+        patron_type: str = fields["patronType"].get("key", "")
+
+        patrondata.external_type = patron_type
+
+        # Basic block reasons
+
+        # Does the patron type start with the library prefix
+        # This ensures the patron is a member of the library
+        # they are interacting with
+        if not patron_type.startswith(self.sirsi_library_prefix):
+            patrondata.block_reason = SirsiBlockReasons.INCORRECT_LOCATION
```

Comment:
I'm concerned with these being blocks, since blocks are enforced when a patron goes to actually checkout the content, not when they authenticate. Settings these as blocks will mean other library patrons will seem like they can log in, until they take an action in the library.

I'm also concerned because we have overrides for libraries that want to allow patrons to checkout content despite blocks, which would override these blocks getting set.

https://github.com/ThePalaceProject/circulation/blob/03fe8e5d02bce41d3d68147d29ff0d8bd30b61af/api/authenticator.py#L2095-L2197

Reply:
Moved the prefix check to the "enforce" method

Reply:
We will always have the 12 hours gap in any restriction changes for the patron due to the needs_external_sync gap though